src/core/cpu_patches.cpp

```diff
@@ -0,0 +1,381 @@
+    ASSERT_MSG(operand.type == ZYDIS_OPERAND_TYPE_MEMORY, "Expected memory operand, got type: {}",
+               static_cast<u32>(operand.type));
+
+    Xbyak::RegExp expression{};
+    if (operand.mem.base != ZYDIS_REGISTER_NONE) {
+        expression = expression + ZydisToXbyakRegister(operand.mem.base);
+    }
+    if (operand.mem.index != ZYDIS_REGISTER_NONE) {
+        if (operand.mem.scale != 0) {
+            expression = expression + ZydisToXbyakRegister(operand.mem.index) * operand.mem.scale;
+        } else {
+            expression = expression + ZydisToXbyakRegister(operand.mem.index);
+        }
+    }
+    if (operand.mem.disp.size != 0 && operand.mem.disp.value != 0) {
+        expression = expression + operand.mem.disp.value;
+    }
+
+    return Xbyak::util::ptr[expression];
+}
+
+static std::unique_ptr<Xbyak::Operand> ZydisToXbyakOperand(const ZydisDecodedOperand& operand) {
+    switch (operand.type) {
+    case ZYDIS_OPERAND_TYPE_REGISTER: {
+        return std::make_unique<Xbyak::Reg>(ZydisToXbyakRegisterOperand(operand));
+    }
+    case ZYDIS_OPERAND_TYPE_MEMORY: {
+        return std::make_unique<Xbyak::Address>(ZydisToXbyakMemoryOperand(operand));
+    }
+    default:
+        UNREACHABLE_MSG("Unsupported operand type: {}", static_cast<u32>(operand.type));
+    }
+}
+
+static bool OperandUsesRegister(const Xbyak::Operand* operand, int index) {
+    if (operand->isREG()) {
+        return operand->getIdx() == index;
+    }
+    if (operand->isMEM()) {
+        const Xbyak::RegExp& reg_exp = operand->getAddress().getRegExp();
+        return reg_exp.getBase().getIdx() == index || reg_exp.getIndex().getIdx() == index;
+    }
+    UNREACHABLE_MSG("Unsupported operand kind: {}", static_cast<u32>(operand->getKind()));
+}
+
+static bool IsRegisterAllocated(
+    const std::initializer_list<const Xbyak::Operand*>& allocated_registers, const int index) {
+    return std::ranges::find_if(allocated_registers.begin(), allocated_registers.end(),
+                                [index](const Xbyak::Operand* operand) {
+                                    return OperandUsesRegister(operand, index);
+                                }) != allocated_registers.end();
+}
+
+static Xbyak::Reg AllocateScratchRegister(
+    const std::initializer_list<const Xbyak::Operand*> allocated_registers, const u32 bits) {
+    for (int index = Xbyak::Operand::R8; index <= Xbyak::Operand::R15; index++) {
+        if (!IsRegisterAllocated(allocated_registers, index)) {
+            return Xbyak::Reg32e(index, static_cast<int>(bits));
+        }
+    }
+    UNREACHABLE_MSG("Out of scratch registers!");
+}
+
+static bool GenerateTlsAccess(const ZydisDecodedOperand* operands, Xbyak::CodeGenerator& c) {
+    const auto& dst_op = operands[0];
+    const auto& src_op = operands[1];
+    if (src_op.type != ZYDIS_OPERAND_TYPE_MEMORY || src_op.mem.segment != ZYDIS_REGISTER_FS ||
```

Comment:
Will this also account for 64-bit TLS access pattern

Reply:
It accounts for `mov (64-bit register), fs:[0]` with any size for the zero, 32 or 64 bit. We could support non-zero offset as well if needed by removing some of the checks here, but I kept it in line with the current support. Especially since the more specific we get here, the better it guards against false-read instructions from data in between code.